hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/ExchangeRateManager.java

```diff
@@ -45,25 +60,40 @@
  */
 @Singleton
 public final class ExchangeRateManager {
-    private static final Logger logger = LogManager.getLogger(ExchangeRateManager.class);
-    private static final ExchangeRateSet DEFAULT_EXCHANGE_RATES = ExchangeRateSet.DEFAULT;
-
-    /** The parsed {@link ExchangeRateSet}. This is placed, in its entirety, into the receipt for a transaction. */
-    private ExchangeRateSet exchangeRates;
-    /** The "current" rate. This is from the {@link #exchangeRates}, but sanitized in case of bad input. */
-    private ExchangeRate currentRate;
-    /** The "next" rate. This is from the {@link #exchangeRates}, but sanitized in case of bad input. */
-    private ExchangeRate nextRate;
-    /** The expiration time of the "current" rate, in consensus seconds since the epoch, but cached for speed. */
-    private long currentRateExpirationSeconds;
+
+    private static final BigInteger ONE_HUNDRED = BigInteger.valueOf(100);
+
+    private final ConfigProvider configProvider;
+
+    private ExchangeRateInfo currentExchangeRateInfo;
+    private ExchangeRateSet midnightRates;
 
     @Inject
-    public ExchangeRateManager() {
-        // Initialize the exchange rate set. The default is not particularly useful, but isn't null.
-        this.exchangeRates = DEFAULT_EXCHANGE_RATES;
-        this.currentRate = exchangeRates.currentRateOrElse(ExchangeRate.DEFAULT);
-        this.nextRate = exchangeRates.nextRateOrElse(currentRate);
-        this.currentRateExpirationSeconds = 0;
+    public ExchangeRateManager(@NonNull final ConfigProvider configProvider) {
+        this.configProvider = requireNonNull(configProvider, "configProvider must not be null");
+    }
+
+    public void init(@NonNull final HederaState state, @NonNull final Bytes bytes) {
+        requireNonNull(state, "state must not be null");
+        requireNonNull(bytes, "bytes must not be null");
+
+        // First we try to read midnightRates from state
+        midnightRates = state.createReadableStates(FeeService.NAME)
+                .<ExchangeRateSet>getSingleton(FeeService.MIDNIGHT_RATES_STATE_KEY)
+                .get();
+        if (midnightRates != ExchangeRateSet.DEFAULT) {
+            // midnightRates were found in state, a regular update is sufficient
+            update(bytes);
+        }
+
+        // If midnightRates were not found in state, we initialize them from the file
+        try {
+            midnightRates = ExchangeRateSet.PROTOBUF.parse(bytes.toReadableSequentialData());
+        } catch (IOException e) {
+            // an error here is fatal and needs to be handled by the general initialization code
+            throw new UncheckedIOException(e);
```

Comment:
Would it be worth a logging statement here as well to indicate what went wrong?

Reply:
The `UncheckedIOException` will be caught somewhere and logged. I hate it when applications log the same exception multiple times. It always confuses me. But I will add a message to the `UncheckedIOException`.